### PR TITLE
Add sortBy/sortDirection for favorite retrieval

### PR DESCRIPTION
**Pull request by nickgros**

### CODE REVIEW

File: lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/favorite/SortBy.json [hunk 1]
```diff
@@ -0,0 +1,11 @@
+{
+	"type": "string",
+	"enum": [{
+		"name": "FAVORITED_ON",
+		"description": "TODO: Auto-generated description"
```
Comment (brucehoff) [inline lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/favorite/SortBy.json]: fill in the two descriptions, please

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/UserProfileManager.java [hunk 2]
```diff
@@ -90,7 +91,7 @@ public void removeFavorite(UserInfo userInfo, String entityId)
 	 * Retrieve users list of favorites, paginated
 	 */
 	public PaginatedResults<EntityHeader> getFavorites(UserInfo userInfo,
-			int limit, int offset) throws DatastoreException,
+													   int limit, int offset, SortBy sortBy, org.sagebionetworks.repo.model.favorite.SortDirection sortDirection) throws DatastoreException,
```
Comment (brucehoff) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/UserProfileManager.java]: Why such ginormous indentation?

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/UserProfileManagerImpl.java [hunk 3]
```diff
@@ -210,12 +211,12 @@ public void removeFavorite(UserInfo userInfo, String entityId)
 
 	@Override
 	public PaginatedResults<EntityHeader> getFavorites(UserInfo userInfo,
-			int limit, int offset) throws DatastoreException,
+													   int limit, int offset, SortBy sortBy, org.sagebionetworks.repo.model.favorite.SortDirection sortDirection) throws DatastoreException,
```
Comment (brucehoff) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/UserProfileManagerImpl.java]: ginormous indentation

File: services/repository/src/main/java/org/sagebionetworks/repo/web/controller/UserProfileController.java [hunk 4]
```diff
@@ -444,9 +445,15 @@ public void removeFavorite(
 	PaginatedResults<EntityHeader> getFavorites(
 			@RequestParam(value = AuthorizationConstants.USER_ID_PARAM) Long userId,
 			@RequestParam(value = ServiceConstants.PAGINATION_OFFSET_PARAM, required = false, defaultValue = ServiceConstants.DEFAULT_PAGINATION_OFFSET_PARAM) Integer offset,
-			@RequestParam(value = ServiceConstants.PAGINATION_LIMIT_PARAM, required = false, defaultValue = ServiceConstants.DEFAULT_PAGINATION_LIMIT_PARAM) Integer limit) 
+			@RequestParam(value = ServiceConstants.PAGINATION_LIMIT_PARAM, required = false, defaultValue = ServiceConstants.DEFAULT_PAGINATION_LIMIT_PARAM) Integer limit,
+			@RequestParam(value = ServiceConstants.SORT_BY_PARAM, required = false, defaultValue = "FAVORITED_ON") String sortBy,
```
Comment (brucehoff) [inline services/repository/src/main/java/org/sagebionetworks/repo/web/controller/UserProfileController.java]: seems like the strings FAVORITED_ON and DESC are defined in `org.sagebionetworks.repo.model.favorite.SortBy` and `org.sagebionetworks.repo.model.favorite.SortDirection`.  Can you use those values?

File: services/repository/src/main/java/org/sagebionetworks/repo/web/service/UserProfileServiceImpl.java [hunk 5]
```diff
@@ -219,10 +220,10 @@ public void removeFavorite(Long userId, String entityId)
 
 	@Override
 	public PaginatedResults<EntityHeader> getFavorites(Long userId, int limit,
-			int offset) throws DatastoreException, InvalidModelException,
+                                                       int offset, SortBy sortBy, org.sagebionetworks.repo.model.favorite.SortDirection sortDirection) throws DatastoreException, InvalidModelException,
```
Comment (brucehoff) [inline services/repository/src/main/java/org/sagebionetworks/repo/web/service/UserProfileServiceImpl.java]: ginormous indentation?

File: lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/dao/DBOFavoriteDAOImplAutowiredTest.java [hunk 6]
```diff
@@ -263,7 +265,7 @@ public void testGetFavoritesEntityHeader() throws Exception {
 		Favorite fav2created = favoriteDao.add(fav2);
 		favoritesToDelete.add(fav2created);
 
-		PaginatedResults<EntityHeader> favs = favoriteDao.getFavoritesEntityHeader(creatorUserGroupId.toString(), Integer.MAX_VALUE, 0);
+		PaginatedResults<EntityHeader> favs = favoriteDao.getFavoritesEntityHeader(creatorUserGroupId.toString(), Integer.MAX_VALUE, 0, SortBy.FAVORITED_ON, SortDirection.ASC);
```
Comment (brucehoff) [inline lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/dao/DBOFavoriteDAOImplAutowiredTest.java:272]: How do you know that fav2created will have a later timestamp than fav1created?  They could be created in the same millisecond, have the same time stamp, and then the outcome of this test becomes random.  Either mock the clock in the DAO or put a > 1 ms delay between creation of the two favorites.

**Review by brucehoff — CHANGES_REQUESTED** [review]

File: services/repository/src/main/java/org/sagebionetworks/repo/web/controller/UserProfileController.java (hunk 4 above)
Comment (nickgros) [reply services/repository/src/main/java/org/sagebionetworks/repo/web/controller/UserProfileController.java]: This is unfortunately not possible: https://stackoverflow.com/a/3272859

Instead I defined the strings in ServiceConstants and reference them here.

File: lib/lib-shared-models/src/main/java/org/sagebionetworks/repo/model/ServiceConstants.java [hunk 7]
```diff
@@ -1,5 +1,7 @@
 package org.sagebionetworks.repo.model;
 
+import org.sagebionetworks.repo.model.favorite.SortBy;
+
```
Comment (brucehoff) [inline lib/lib-shared-models/src/main/java/org/sagebionetworks/repo/model/ServiceConstants.java]: Why is this import added?

File: services/repository/src/main/java/org/sagebionetworks/repo/web/controller/UserProfileController.java [hunk 8]
```diff
@@ -65,6 +66,9 @@
 @RequestMapping(UrlHelpers.REPO_PATH)
 public class UserProfileController {
 
+	private static final String SORT_BY_FAVORITES_DEFAULT = SortBy.FAVORITED_ON.toString();
+	private static final String SORT_DIRECTION_FAVORITES_DEFAULT = org.sagebionetworks.repo.model.favorite.SortDirection.DESC.toString();
```
Comment (brucehoff) [inline services/repository/src/main/java/org/sagebionetworks/repo/web/controller/UserProfileController.java]: Where are these new constants used?

**Review by brucehoff — CHANGES_REQUESTED** [review]

File: services/repository/src/main/java/org/sagebionetworks/repo/web/controller/UserProfileController.java (hunk 8 above)
Comment (nickgros) [reply services/repository/src/main/java/org/sagebionetworks/repo/web/controller/UserProfileController.java]: Nowhere, I forgot to delete them

File: lib/lib-shared-models/src/main/java/org/sagebionetworks/repo/model/ServiceConstants.java (hunk 7 above)
Comment (nickgros) [reply lib/lib-shared-models/src/main/java/org/sagebionetworks/repo/model/ServiceConstants.java]: I was trying to create a string constant using the enum and forgot to remove the import after I realized that I couldn't use it.

**Review by brucehoff — APPROVED** [review]
